keps/sig-auth/20191101-add-request-id-to-k8s-logs.md

```diff
@@ -0,0 +1,308 @@
+| pkg/kubelet/volumemanager/reconciler/reconciler.go | unmountDetachDevices |
+| pkg/kubelet/remote/remote_runtime.go | ContainerStatus |
+| pkg/volume/secret/secret.go | SetUpAt |
+| pkg/kubelet/status/status_manager.go | updateStatusInternal |
+| pkg/kubelet/status/status_manager.go | syncPod |
+| pkg/volume/util/util.go | UnmountViaEmptyDir |
+| pkg/kubelet/volumemanager/volume_manager.go | WaitForAttachAndMount |
+
+**kube-proxy**
+
+```
+NONE
+```
+
+**kube-scheduler**
+
+| source file | function name |
+| ------ | ------ |
+| pkg/scheduler/framework/plugins/defaultbinder/default_binder.go | Bind |
+| pkg/scheduler/eventhandlers.go | addPodToSchedulingQueue |
+| pkg/scheduler/eventhandlers.go | updatePodInSchedulingQueue |
+| pkg/scheduler/eventhandlers.go | deletePodFromSchedulingQueue |
+| pkg/scheduler/eventhandlers.go | addPodToCache |
+| pkg/scheduler/eventhandlers.go | updatePodInCache |
+| pkg/scheduler/eventhandlers.go | deletePodFromCache |
+| pkg/scheduler/scheduler.go | scheduleOne |
+
+**NOTE1:**
+
+We need context or object or http request to get Trace-ID(This is specification of `Distributed Tracing` feature). So, we may exclude function which does not have such resource(context, object, http request) from our target.
+
+**NOTE2:**
+
+We use Tracing codes by `Distributed Tracing` KEP for propagation. So the following implementation is required.
+ - Case1. The function which contains Tracing codes by `Distributed Tracing` KEP
+   - We just add Request-ID codes (Request-ID codes don't interfere Tracing codes).
+ - Case2. The function which does not contain Tracing codes by `Distributed Tracing` KEP
+   - We add both of Tracing codes and Request-id codes (Request-ID codes don't interfere Tracing codes).
+
+#### Design of Export Request-ID
+
+We get Trace-ID by using `Distributed Tracing` feature in each Kubernetes function that we add Tracing codes. And then, we add Trace-ID information into **each klog call** as Request-ID. Note that we don't associate Request-ID to all of klog calls. Our target is important operations such as `kubectl create/delete/etc.`, and our target klog calls are the only klogs which is called via such important operations. Request-ID feature does not change existing klog function/method, but changes each klog calls and their log format. Currently, there is [Structured logging](https://github.com/serathius/enhancements/blob/structured-logging/keps/sig-instrumentation/20191115-structured-logging.md) KEP, and this KEP also change specific klog calls. We will merge Request-ID feature after Structured logging KEP is merged.
+
+**Examples (We quote some parts from `Structured logging` KEP)**
+
+Source Code
+
+Original
+```go
+klog.Infof("Updated pod %s status to ready", pod.name)
+```
+
+Structured format
+```go
+klog.InfoS("Pod status updated", "pod", pod, "status", "ready")
+```
+Request-ID with structured format
+```go
+klog.InfoS("Request-ID", trace-id, "Pod status updated", "pod", pod, "status", "ready")
```

Comment:
"Request-ID should be a normal field, meaning it should be put after message "Pod status updated"

Reply:
OK, I put request-id after the message.
btw, what is normal field? (sorry, I'm not familiar the structure of field of klog in k8s...)